cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -0,0 +1,674 @@
+    this.celCompiler = CelCompilerFactory.standardCelCompilerBuilder().setOptions(options).build();
+  }
+
+  /**
+   * Generates a Client-Side CAB token given the {@link CredentialAccessBoundary}.
+   *
+   * @param accessBoundary The credential access boundary that defines the restrictions for the
+   *     generated CAB token.
+   * @return The Client-Side CAB token in an {@link AccessToken} object
+   * @throws IOException If an I/O error occurs while refreshing the source credentials
+   * @throws CelValidationException If the availability condition is an invalid CEL expression
+   * @throws GeneralSecurityException If an error occurs during encryption
+   */
+  public AccessToken generateToken(CredentialAccessBoundary accessBoundary)
+      throws IOException, CelValidationException, GeneralSecurityException {
+    this.refreshCredentialsIfRequired();
+
+    String intermediateToken, sessionKey;
+    Date intermediateTokenExpirationTime;
+
+    synchronized (refreshLock) {
+      intermediateToken = this.intermediateCredentials.intermediateAccessToken.getTokenValue();
+      intermediateTokenExpirationTime =
+          this.intermediateCredentials.intermediateAccessToken.getExpirationTime();
+      sessionKey = this.intermediateCredentials.accessBoundarySessionKey;
+    }
+
+    byte[] rawRestrictions = this.serializeCredentialAccessBoundary(accessBoundary);
+
+    byte[] encryptedRestrictions = this.encryptRestrictions(rawRestrictions, sessionKey);
+
+    String tokenValue =
+        intermediateToken + "." + Base64.getUrlEncoder().encodeToString(encryptedRestrictions);
+
+    return new AccessToken(tokenValue, intermediateTokenExpirationTime);
+  }
+
+  /**
+   * Refreshes the intermediate access token and access boundary session key if required.
+   *
+   * <p>This method checks the expiration time of the current intermediate access token and
+   * initiates a refresh if necessary. The refresh process also refreshes the underlying source
+   * credentials.
+   *
+   * @throws IOException If an error occurs during the refresh process, such as network issues,
+   *     invalid credentials, or problems with the token exchange endpoint.
+   */
+  @VisibleForTesting
+  void refreshCredentialsIfRequired() throws IOException {
+    RefreshType refreshType = determineRefreshType();
+
+    if (refreshType == RefreshType.NONE) {
+      return; // No refresh needed, token is still valid.
+    }
+
+    // If a refresh is required, create or retrieve the refresh task.
+    RefreshTask refreshTask = getOrCreateRefreshTask();
+
+    // Handle the refresh based on the determined refresh type.
+    switch (refreshType) {
+      case BLOCKING:
+        if (refreshTask.isNew) {
+          // Start a new refresh task only if the task is new
+          MoreExecutors.directExecutor().execute(refreshTask.task);
+        }
+        try {
+          refreshTask.task.get(); // Wait for the refresh task to complete.
```

Comment:
nit: Let's be consistent with comments (also change elsewhere)

```suggestion
          // Wait for the refresh task to complete.
          refreshTask.task.get();
```